modules/custom/wim_cvdr/wim_cvdr.module

```diff
@@ -0,0 +1,235 @@
+<?php
+/**
+ * @file
+ * WIM CVDR main module file.
+ */
+
+/**
+ * Implements hook_menu().
+ */
+function wim_cvdr_menu() {
+  $items = array();
+
+  $items['admin/config/services/cvdr'] = array(
+    'title' => 'WIM CVDR',
+    'description' => 'Configuration for connecting with CVDR.',
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('_wim_cvdr_admin_page'),
+    'access arguments' => array('administer blocks'),
+    'type' => MENU_NORMAL_ITEM,
+    'file' => 'wim_cvdr.admin.inc',
+  );
+
+  return $items;
+}
+
+/**
+ * Implements hook_block_info().
+ */
+function wim_cvdr_block_info() {
+  $blocks = array();
+  $subjects = _wim_cvdr_retrieve_subjects();
+
+  // Create a seperate block for each subject.
+  foreach ($subjects as $subject) {
+    // Create nice machine name based on subject name.
+    $machine_name = str_replace('-', '_', drupal_html_class($subject));
+    $blocks[md5('wim_cvdr_' . $machine_name)] = array(
+      'info' => t('CVDR: !subject', array('!subject' => $subject)),
+      'cache' => DRUPAL_CACHE_GLOBAL,
+    );
+  }
+
+  return $blocks;
+}
+
+/**
+ * Implements hook_block_view().
+ */
+function wim_cvdr_block_view($delta = '') {
+  $block = array();
+
+  $subjects = _wim_cvdr_retrieve_subjects();
+  foreach ($subjects as $subject) {
+    // Create nice machine name based on subject name.
+    $machine_name = str_replace('-', '_', drupal_html_class($subject));
+
+    // Check if block is current subject.
+    if ($delta === md5('wim_cvdr_' . $machine_name)) {
+      // Render subject block.
+      $block['subject'] = $subject;
+
+      // Build the links to the CVDR items.
+      $items = _wim_cvdr_retrieve_items($subject);
+
+      // Sort the items alphabetically.
+      sort($items);
+
+      foreach ($items as $item) {
+        $list_items[] = l($item['title'], $item['url'],
+          array('attributes' => array('target' => '_blank')));
+      }
+
+      $block['content'] = theme('item_list', array(
+        'items' => $list_items,
+        'title' => '',
+        'type' => 'ul',
+        'attributes' => array(
+          'class' => array('cvdr-wim', drupal_html_class($subject)),
+        ),
+      ));
+    }
+  }
+
+  // Add JS.
+  drupal_add_js(array(
+    'wim_cvdr' => array(
+      'maxItems' => variable_get('wim_cvdr_max_visible', 5) - 1,
+    ),
+  ), 'setting');
+  drupal_add_js(
+    drupal_get_path('module', 'wim_cvdr') . '/wim_cvdr.js', array(
+      'type' => 'file',
+      'scope' => 'footer',
+    )
+  );
+
+  return $block;
+}
+
+/**
+ * Retrieve the subjects available from the webservice.
+ *
+ * @return array
+ *   containing the subjects.
+ */
+function _wim_cvdr_retrieve_subjects() {
+  $items = array();
+
+  // Use caching to improve performance.
```

Comment:
This piece of code is also duplicated in _wim_cvdr_retrieve_items

We should try to extract it into a separate function that calls the read_xml and process_xml functions. That function should be responsible for loading from the cache as well as setting the cache.

By moving the caching to a separate function we could in the future call read_xml/process_xml separately without using or updating the cache if this is needed. It's also clearer for future developers where the cached value comes from.